File: integration_tests/python-aiosqlite/test_integration.py
"""Integration tests for scythe-generated aiosqlite queries."""

import asyncio
import os
import sys
from decimal import Decimal
from pathlib import Path

import aiosqlite

from generated.queries import (
    create_order,
    create_user,
    delete_orders_by_user,
    delete_user,
    get_orders_by_user,
    get_user_by_id,
    list_active_users,
)


SCHEMA_PATH = Path(__file__).parent.parent / "sql" / "sqlite" / "schema.sql"


def get_database_url() -> str:
    """Read SQLITE_PATH from environment."""
    url = os.environ.get("SQLITE_PATH")
    if not url:
        print("ERROR: SQLITE_PATH environment variable is not set", file=sys.stderr)
        sys.exit(1)
    return url


async def setup_schema(conn: aiosqlite.Connection) -> None:
    """Drop all tables and recreate schema from SQL file."""
    await conn.execute("DROP TABLE IF EXISTS user_tags")
    await conn.execute("DROP TABLE IF EXISTS tags")
    await conn.execute("DROP TABLE IF EXISTS orders")
    await conn.execute("DROP TABLE IF EXISTS users")
    schema_sql = SCHEMA_PATH.read_text()
    await conn.executescript(schema_sql)
    await conn.commit()


async def test_create_user(conn) -> int:
    """Test CreateUser query. Returns created user ID."""
    await create_user(conn, name="Alice", email="alice@example.com", status="active")
    async with conn.cursor() as cur:
        await cur.execute("SELECT last_insert_rowid()")
        row = await cur.fetchone()
        user_id = row[0]
    user = await get_user_by_id(conn, id=user_id)
    assert user is not None, "CreateUser returned None"
    assert user.name == "Alice", f"Expected name 'Alice', got '{user.name}'"
    assert user.email == "alice@example.com", f"Expected email 'alice@example.com', got '{user.email}'"
    print("PASS: CreateUser")
    return user.id


async def test_get_user_by_id(conn, user_id: int) -> None:
    """Test GetUserById query."""
    user = await get_user_by_id(conn, id=user_id)
    assert user is not None, f"GetUserById returned None for id={user_id}"
    assert user.name == "Alice", f"Expected name 'Alice', got '{user.name}'"
    assert user.id == user_id, f"Expected id {user_id}, got {user.id}"
    print("PASS: GetUserById")


async def test_list_active_users(conn) -> None:
    """Test ListActiveUsers query."""
    users = await list_active_users(conn, status="active")
    assert len(users) >= 1, f"Expected at least 1 active user, got {len(users)}"
    names = [u.name for u in users]
    assert "Alice" in names, f"Expected 'Alice' in active users, got {names}"
    print("PASS: ListActiveUsers")


async def test_create_order(conn, user_id: int) -> int:
    """Test CreateOrder query. Returns created order ID."""
    await create_order(conn, user_id=user_id, total=49.99, notes="Test order")
    orders = await get_orders_by_user(conn, user_id=user_id)
    order = orders[0] if orders else None
    assert order is not None, "CreateOrder returned None"
    assert order.notes == "Test order", f"Expected notes 'Test order', got '{order.notes}'"
    print("PASS: CreateOrder")
    return order.id


async def test_get_orders_by_user(conn, user_id: int) -> None:
    """Test GetOrdersByUser query."""
    orders = await get_orders_by_user(conn, user_id=user_id)
    assert len(orders) >= 1, f"Expected at least 1 order, got {len(orders)}"
    assert orders[0].notes == "Test order", f"Expected notes 'Test order', got '{orders[0].notes}'"
    print("PASS: GetOrdersByUser")


async def test_delete_user(conn, user_id: int) -> None:
    """Test DeleteUser query."""
    # Delete orders first due to FK constraint
    await delete_orders_by_user(conn, user_id=user_id)
    await delete_user(conn, id=user_id)
    user = await get_user_by_id(conn, id=user_id)
    assert user is None, f"Expected user to be deleted, but got {user}"
    print("PASS: DeleteUser")


async def run_tests() -> None:
    """Run all integration tests."""
    database_url = get_database_url()
    conn = await aiosqlite.connect(database_url)
    try:
        await setup_schema(conn)

        user_id = await test_create_user(conn)
        await test_get_user_by_id(conn, user_id)
        await test_list_active_users(conn)
        order_id = await test_create_order(conn, user_id)
        await test_get_orders_by_user(conn, user_id)
        await test_delete_user(conn, user_id)
    finally:
        await conn.close()

    print("\nALL TESTS PASSED")


if __name__ == "__main__":
    try:
        asyncio.run(run_tests())
    except Exception as exc:
        print(f"FAIL: {exc}", file=sys.stderr)
        sys.exit(1)

File: integration_tests/php-pdo-sqlite/test_integration.php
<?php

declare(strict_types=1);

if (file_exists(__DIR__ . '/vendor/autoload.php')) {
    require_once __DIR__ . '/vendor/autoload.php';
}
require_once __DIR__ . '/generated/queries.php';

use App\Generated\Queries;
use App\Generated\RecordNotFoundException;
use App\Generated\CreateUserRow;
use App\Generated\GetUserByIdRow;
use App\Generated\ListActiveUsersRow;
use App\Generated\CreateOrderRow;
use App\Generated\GetOrdersByUserRow;

function get_database_url(): string
{
    $url = getenv('SQLITE_PATH');
    if ($url === false || $url === '') {
        fwrite(STDERR, "ERROR: SQLITE_PATH environment variable is not set\n");
        exit(1);
    }
    return $url;
}

/**
 * Splits a SQL script into statements on top-level ';' only -- unlike a
 * naive `explode(';', $sql)`, this tracks single- and double-quoted spans,
 * PostgreSQL dollar-quoted bodies, and '--' line comments (an apostrophe in
 * a comment must not open a phantom string -- board #224 follow-up) so a
 * ';' inside a string literal, a `$$ ... $$` function body, or a comment
 * does not split the statement in half. '/* ... *' + '/' block comments are
 * not handled -- no schema under integration_tests/sql/ uses them today.
 *
 * @return array<string>
 */
function split_sql_statements(string $sql): array
{
    $statements = [];
    $current = '';
    $inSingle = false;
    $inDouble = false;
    $inLineComment = false;
    $dollarTag = null;
    $length = strlen($sql);
    $i = 0;
    while ($i < $length) {
        $ch = $sql[$i];
        if ($inLineComment) {
            $current .= $ch;
            if ($ch === "\n") {
                $inLineComment = false;
            }
            $i++;
            continue;
        }
        if ($dollarTag !== null) {
            $current .= $ch;
            if ($ch === '$' && substr($sql, $i, strlen($dollarTag)) === $dollarTag) {
                $current .= substr($dollarTag, 1);
                $i += strlen($dollarTag);
                $dollarTag = null;
                continue;
            }
            $i++;
            continue;
        }
        if ($inSingle) {
            $current .= $ch;
            if ($ch === "'") {
                $inSingle = false;
            }
            $i++;
            continue;
        }
        if ($inDouble) {
            $current .= $ch;
            if ($ch === '"') {
                $inDouble = false;
            }
            $i++;
            continue;
        }
        if ($ch === "'") {
            $inSingle = true;
            $current .= $ch;
            $i++;
            continue;
        }
        if ($ch === '"') {
            $inDouble = true;
            $current .= $ch;
            $i++;
            continue;
        }
        if ($ch === '-' && ($sql[$i + 1] ?? '') === '-') {
            $inLineComment = true;
            $current .= $ch;
            $i++;
            continue;
        }
        if ($ch === '$' && preg_match('/\G\$[A-Za-z0-9_]*\$/', $sql, $matches, 0, $i) === 1) {
            $dollarTag = $matches[0];
            $current .= $dollarTag;
            $i += strlen($dollarTag);
            continue;
        }
        if ($ch === ';') {
            $statements[] = $current;
            $current = '';
            $i++;
            continue;
        }
        $current .= $ch;
        $i++;
    }
    if (trim($current) !== '') {
        $statements[] = $current;
    }
    return array_values(array_filter(array_map('trim', $statements), static fn (string $stmt): bool => $stmt !== ''));
}



function create_pdo(string $path): PDO
{
    $dsn = sprintf('sqlite:%s', $path);
    $pdo = new PDO($dsn, null, null, [
        PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION,
        PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC,
    ]);
    return $pdo;
}

function setup_schema(PDO $pdo): void
{
    $pdo->exec("DROP TABLE IF EXISTS user_tags");
    $pdo->exec("DROP TABLE IF EXISTS tags");
    $pdo->exec("DROP TABLE IF EXISTS orders");
    $pdo->exec("DROP TABLE IF EXISTS users");
    $schema_path = __DIR__ . '/../sql/sqlite/schema.sql';
    $schema_sql = file_get_contents($schema_path);
    if ($schema_sql === false) {
        throw new RuntimeException("Failed to read schema file: {$schema_path}");
    }
    $pdo->exec($schema_sql);
}

function assert_equal(mixed $expected, mixed $actual, string $message): void
{
    if ($expected !== $actual) {
        throw new RuntimeException(
            "Assertion failed: {$message} (expected " . var_export($expected, true)
            . ", got " . var_export($actual, true) . ")"
        );
    }
}

function assert_not_null(mixed $value, string $message): void
{
    if ($value === null) {
        throw new RuntimeException("Assertion failed: {$message} (got null)");
    }
}

function assert_true(bool $value, string $message): void
{
    if (!$value) {
        throw new RuntimeException("Assertion failed: {$message}");
    }
}

function assert_null(mixed $value, string $message): void
{
    if ($value !== null) {
        throw new RuntimeException("Assertion failed: {$message} (expected null, got " . var_export($value, true) . ")");
    }
}



function test_create_user(PDO $pdo): int
{
    Queries::createUser($pdo, "Alice", "alice@example.com", "active");
    $user = Queries::getUserById($pdo, 1);
    assert_not_null($user, "CreateUser returned null");
    assert_equal("Alice", $user->name, "CreateUser name");
    assert_equal("alice@example.com", $user->email, "CreateUser email");
    echo "PASS: CreateUser\n";
    return $user->id;
}

function test_get_user_by_id(PDO $pdo, int $user_id): void
{
    $user = Queries::getUserById($pdo, $user_id);
    assert_not_null($user, "GetUserById returned null for id={$user_id}");
    assert_equal("Alice", $user->name, "GetUserById name");
    assert_equal($user_id, $user->id, "GetUserById id");
    echo "PASS: GetUserById\n";
}

function test_list_active_users(PDO $pdo): void
{
    $users = iterator_to_array(Queries::listActiveUsers($pdo, "active"));
    assert_true(count($users) >= 1, "Expected at least 1 active user, got " . count($users));
    $names = array_map(fn($u) => $u->name, $users);
    assert_true(in_array("Alice", $names, true), "Expected 'Alice' in active users");
    echo "PASS: ListActiveUsers\n";
}

function test_create_order(PDO $pdo, int $user_id): int
{
    Queries::createOrder($pdo, $user_id, 49.99, "Test order");
    $orders = iterator_to_array(Queries::getOrdersByUser($pdo, $user_id));
    $order = !empty($orders) ? $orders[0] : null;
    assert_not_null($order, "CreateOrder returned null");
    assert_equal("Test order", $order->notes, "CreateOrder notes");
    echo "PASS: CreateOrder\n";
    return $order->id;
}

function test_get_orders_by_user(PDO $pdo, int $user_id, int $order_id): void
{
    $orders = iterator_to_array(Queries::getOrdersByUser($pdo, $user_id));
    assert_true(count($orders) >= 1, "Expected at least 1 order, got " . count($orders));
    assert_equal("Test order", $orders[0]->notes, "GetOrdersByUser notes");
    $found_ids = array_map(fn($o) => $o->id, $orders);
    assert_true(in_array($order_id, $found_ids, true), "Expected order $order_id in results, got " . implode(", ", $found_ids));
    echo "PASS: GetOrdersByUser\n";
}

function test_delete_user(PDO $pdo, int $user_id): void
{
    // Delete orders first due to FK constraint
    Queries::deleteOrdersByUser($pdo, $user_id);
    Queries::deleteUser($pdo, $user_id);
    // getUserById is `:one`, so a missing row throws RecordNotFoundException rather than
    // returning null.
    try {
        Queries::getUserById($pdo, $user_id);
        throw new RuntimeException("Expected getUserById to throw RecordNotFoundException, but it returned a row");
    } catch (RecordNotFoundException) {
        // expected: the user was deleted
    }
    echo "PASS: DeleteUser\n";
}

try {
    $database_url = get_database_url();
    $pdo = create_pdo($database_url);

    setup_schema($pdo);

    $user_id = test_create_user($pdo);
    test_get_user_by_id($pdo, $user_id);
    test_list_active_users($pdo);
    $order_id = test_create_order($pdo, $user_id);
    test_get_orders_by_user($pdo, $user_id, $order_id);
    test_delete_user($pdo, $user_id);

    echo "\nALL TESTS PASSED\n";
    exit(0);
} catch (Throwable $e) {
    fwrite(STDERR, "FAIL: " . $e->getMessage() . "\n");
    fwrite(STDERR, $e->getTraceAsString() . "\n");
    exit(1);
}
